### DISCUSSION TITLE
Feature: Allow shadowing cipherdir with mountpoint

### DISCUSSION THREAD

**Discussion by Atrate** [rfjakob/gocryptfs#718] [Ideas & Feature requests]
It would be nice if `gocryptfs` allowed shadowing the cipherdir with its decrypted mountpoint, like `ecryptfs`.

I am not knowledgeable enough as to whether FUSE supports it or whether it's possible to do without superuser rights/a kernel module, so I'm just throwing the idea here.

Current behavior: 
```
> gocryptfs /tmp/tmp.2CrICJxW7V/test /tmp/tmp.2CrICJxW7V/test
Mountpoint "/tmp/tmp.2CrICJxW7V/test" would shadow cipherdir "/tmp/tmp.2CrICJxW7V/test", this is not supported
```

**Comment by rfjakob** [comment]
Hi, this is theoretically possible, but isn't it just super confusing?

**Reply by Atrate** [reply]
I guess I just got used to the way ecryptfs allows it and it became convenient for me to not have to keep track of two separate directories.

**Reply by kaoneko** [reply]
On the contrary, it makes a lot of sense (from an end-user perspective) to point gocryptfs at a directory in order decrypt it, or unlock it, without having to look in another place to view the decrypted content.

I wrote a wrapper script that just takes `CIPHERDIR` as an argument, renames it to `.CIPHERDIR` (hidden folder) and then mounts it on `CIPHERDIR` after recreating the folder. And does the reverse when `CIPHERDIR` is already a fuse.gocryptfs mountpoint. The resulting experience feels very natural.

Just like some apps or operating systems offer a "locked folder" for your most sensitive files, that you gain access to only after authenticating upon opening it; the "unlocking the vault" experience.

**Reply by rfjakob** [reply]
Doesn't this create the risk of a user saving unencrypted files into CIPHERDIR?

**Reply by Sandeep856** [reply]
For forward mode, since the `CIPHERDIR` is now the new mountpoint saving unencrypted files will encrypt the files to the `.CIPHERDIR`. After unmounting revert the changes we made initially that is remove the mountpoint (`CIPHERDIR`) and rename the `.CIPHERDIR` to `CIPHERDIR`

**Comment by Sandeep856** [comment]
I’m facing the same requirement of enabling a shadow mount for a cipher directory. I tried temporarily bypassing the shadow mount check mentioned above, but that resulted in numerous panics.

I’d like to better understand the complexity involved in properly supporting this feature.

**Reply by rfjakob** [reply]
You'd just need to chdir into CIPHERDIR before mounting and then use relative paths everywhere

**Reply by Sandeep856** [reply]
Thanks for the reply, if I understand correctly do you mean something
1. First cd CIPHERDIR
2. Then run gocryptfs . ../CIPHERDIR
Please let me know if this a right approach or would this require changes in the codebase?

**Reply by rfjakob** [reply]
This needs code changes (a lot of code changes)
